cctbx/cctbx_project
# Update conda environments (Python 3.7 and 3.8)

Test updated dependencies and add Python 3.7 (and maybe 3.8) support

## List of changes
1) Add `cctbxlite` environments (not yet enabled)
2) Add `pyside2`, `pyzmq` and `websocket-server` for @Oeffner to `cctbx` environments 
2) Versioned `conda` environments for linux, macOS, and Windows for Python 2.7, 3.6, 3.7, and 3.8
3) Updated CI build to build all versions of Python (12 builds)
a) Windows
b) CentOS 6
c) macOS 10.14 (Xcode 11.3.1)
4) Update Full build to build all versions of Python (40 builds)
a) Windows
b) CentOS 6, 7, 8
c) Ubuntu 14.04, 16.04, 18.04 (will replace 14.04 with 20.04 when it's released later in April)
d) macOS 10.14 Xcode 11.3.1 and 10.3, and macOS 10.15 Xcode 11.4
5) Update Boost to 1.70 (source build), closes #70 
a) Match current version for `conda`
b) Enables Python 3.7 and later support
c) Enables support for latest versions of CUDA
6) Bump macOS minimum deployment version from 10.7 to 10.9
a) Match version for `conda` packages
7) Add `--python` flag to `bootstrap.py` for selecting Python version
a) Just passes argument to `libtbx.install_conda`
8) `scons` moved into `conda` environment
9) Add `--no-boost-src` flag to `bootstrap.py` to prevent downloading of Boost into modules during the `update` step.
10) Add weekly builds of where `-ffast-math` (`/fp:fast` for Windows) is removed and `-O0`, `-O1`, `-O2`, and `-O3` (`/O1` and `/O2` for Windows) are tested on the Full build (160 builds).

@Oeffner Can you double check the Windows builds are working for you?
@luc-j-bourhis @pcxod @Oeffner Are there any changes that you would like to the `cctbxlite` environments? Currently, these are the dependencies in `cctbxlite`
```
biopython
cython
docutils
future
ipython
libsvm
libsvm_py
matplotlib-base
mrcfile
nomkl
numpy
numpydoc
openblas
pillow
pip
psutil
pyopengl
pytest
reportlab
six
sphinx
wxpython
python.app # [osx]
py2app  # [osx]
```
We can probably cut down the number of dependencies for `cctbxlite`.
